sql/catalyst/src/main/scala/org/apache/spark/sql/types/DecimalType.scala

```diff
@@ -148,13 +152,28 @@ object DecimalType extends AbstractDataType {
   }
 
   private[sql] def fromBigDecimal(d: BigDecimal): DecimalType = {
-    DecimalType(Math.max(d.precision, d.scale), d.scale)
+    fromJavaBigDecimal(d.underlying())
+  }
+
+  private[sql] def fromJavaBigDecimal(d: JavaBigDecimal): DecimalType = {
+    val (precision, scale) = if (d.scale < 0 && SQLConf.get.ansiEnabled) {
+      (d.precision - d.scale, 0)
+    } else {
```

Comment:
shall we also handle `precision < scale` like https://github.com/apache/spark/pull/26881/files#diff-92db6fd7dbad00cec24c5878a8c354d9R134 ?

Reply:
added  9d41ea2